From 1e819fc4b09afca2bf2ff13af5b001f6201f419a Mon Sep 17 00:00:00 2001
From: Balazs Scheidler <balazs.scheidler@oneidentity.com>
Date: Tue, 18 Dec 2018 09:53:26 +0100
Subject: [PATCH] cfg-lexer: fix memory leak for dynamically allocated
 CfgBlockGenerators

This should fix a memory leak for app-parser() if that is found in the
configuration.

Signed-off-by: Balazs Scheidler <balazs.scheidler@oneidentity.com>
---
 lib/cfg-lexer.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/lib/cfg-lexer.c b/lib/cfg-lexer.c
index 6b7854f284..020607caa7 100644
--- a/lib/cfg-lexer.c
+++ b/lib/cfg-lexer.c
@@ -731,7 +731,7 @@ _generator_plugin_construct(Plugin *s)
 {
   GeneratorPlugin *self = (GeneratorPlugin *) s;
 
-  return self->gen;
+  return cfg_block_generator_ref(self->gen);
 }
 
 static void
@@ -1023,7 +1023,9 @@ cfg_lexer_preprocess(CfgLexer *self, gint tok, YYSTYPE *yylval, YYLTYPE *yylloc)
       self->cfg &&
       (gen = cfg_lexer_find_generator(self, self->cfg, cfg_lexer_get_context_type(self), yylval->cptr)))
     {
-      if (!cfg_lexer_parse_and_run_block_generator(self, gen, yylval))
+      gboolean success = cfg_lexer_parse_and_run_block_generator(self, gen, yylval);
+      cfg_block_generator_unref(gen);
+      if (!success)
         return CLPR_ERROR;
 
       return CLPR_LEX_AGAIN;
